features/329-principal-apis/design.md

```diff
@@ -0,0 +1,136 @@
+# Background
+Auth gear supports various identity providers, such as password credentials and
+OAuth. Auth gear should support all identity providers equally well, and
+provide a consistent API for manipulating principals.
+
+
+# Use Cases
+- Get which method is used to log in the current session
+- List all methods to authenticated as a user
+- Add/remove user authentication method
+- Send welcome email to OAuth accounts
+- Populate user metadata according to external profile
+- Create password credentials for users signed up with OAuth
+
+
+# Glossary
+
+**Identity Provider**:
+A method to assert the identity of user. For example, providing a password or
+perform OAuth flow.
+
+**Principal**:
+An identity associated with user. A user can have multiple principals. Since
+the term 'Principal' is defined differently across many frameworks and
+standards, we will instead use a simpler term 'Identity' for public interfaces.
+
+
+# Proposed Design
+
+## Re-authentication for Security
+Many principal manipulation operations are critical for security:
+- Add/remove login ID
+- Link/unlink SSO provider
+- Change password
+
+By default, user are required to re-authenticate before performing these
+operations. This is achieve by checking the issue time of the access token is
+recent.
+
+For change password, we support the common case of providing old password to
+implicitly authenticate the user.
+
+Both the re-auth requirement and time to considered 'recent' can be configured.
+Developer can turn off the re-auth requirement in configuration. However,
+developer should understand the potential security risk.
+
+> Related config: `reauthForSecurity` and `reauthIntervalSeconds`
+> Related API: `changePassword` function
+
+## Current Principal
```

Comment:
:s/Principal/Identity/g

Reply:
In this document, we use `Principal`. `Identity` is used for public interface, i.e. in SDK/HTTP API or config.

Reply:
Please only use Principal when appropriate and needed. You're describing a public interface here.